### ISSUE TITLE
PackagePublisher tool fails during retry

### ISSUE THREAD

**Issue by natemcmaster** [aspnet/BuildTools#221]
PackagePublisher fails when it attempts to retry pushing packages. The stack trace seems to indicate this is actually a bug in NuGet. We should investigate if we have used the API incorrectly or if it should be opened as an issue on https://github.com/nuget/home.

http://aspnetci/viewLog.html?buildId=194185&buildTypeId=CoherenceSigned&tab=buildResultsDiv
```
09:39:04]    System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
[09:39:04]       at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
[09:39:04]       at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
[09:39:04]       at System.Net.Http.MultipartContent.Dispose(Boolean disposing)
[09:39:04]       at System.Net.Http.HttpRequestMessage.Dispose(Boolean disposing)
[09:39:04]       at System.Net.Http.HttpRequestMessage.Dispose()
[09:39:04]       at NuGet.Protocol.HttpRetryHandler.<SendAsync>d__0.MoveNext()
[09:39:04]    --- End of stack trace from previous location where exception was thrown ---
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[09:39:04]       at NuGet.Protocol.HttpSource.<GetThrottledResponse>d__16.MoveNext()
[09:39:04]    --- End of stack trace from previous location where exception was thrown ---
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[09:39:04]       at NuGet.Protocol.HttpSource.<ProcessResponseAsync>d__14`1.MoveNext()
[09:39:04]    --- End of stack trace from previous location where exception was thrown ---
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[09:39:04]       at NuGet.Protocol.Core.Types.PackageUpdateResource.<PushPackageToServer>d__19.MoveNext()
[09:39:04]    --- End of stack trace from previous location where exception was thrown ---
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[09:39:04]       at NuGet.Protocol.Core.Types.PackageUpdateResource.<PushPackageCore>d__13.MoveNext()
[09:39:04]    --- End of stack trace from previous location where exception was thrown ---
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[09:39:04]       at NuGet.Protocol.Core.Types.PackageUpdateResource.<PushPackage>d__12.MoveNext()
[09:39:04]    --- End of stack trace from previous location where exception was thrown ---
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[09:39:04]       at NuGet.Protocol.Core.Types.PackageUpdateResource.<Push>d__8.MoveNext()
[09:39:04]    --- End of stack trace from previous location where exception was thrown ---
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[09:39:04]       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[09:39:04]       at PackagePublisher.Program.<PushPackageAsync>d__9.MoveNext() in C:\BuildAgent\work\94c9336e5146215d\src\PackagePublisher\Program.cs:line 171
```

**Comment by pranavkm** [comment]
That stack trace looks like it's coming from here - https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/MultipartContent.cs#L137. Weird

**Comment by pranavkm** [comment]
There's something funky going on here - 

```C#
System.AggregateException: One or more errors occurred. (Adding the specified count to the semaphore would cause it to exceed its maximum count.) ---> System.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.
at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)
at PackagePublisher.Program.<>c__DisplayClass8_1.<<PublishToFeedAsync>b__0>d.MoveNext() in 
```

I wonder if it's because it's an oddity with how directory enumeration works.

**Comment by mikeharder** [comment]
To be clear, this issue is tracking two separate exceptions:

```
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
```

```
System.AggregateException: One or more errors occurred. (Adding the specified count to the semaphore would cause it to exceed its maximum count.) ---> System.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.
```

**Comment by pranavkm** [comment]
For the semaphore bit, we need to look at the result of `WaitAsync` (https://github.com/aspnet/BuildTools/blob/dev/src/PackagePublisher/Program.cs#L147). Probably also bump up the timeout duration.

**Comment by kichalla** [comment]
Fixed with 7cd4d2b05a892563ea2633d5ed8ac5ceace0ada1
